### ISSUE TITLE
Setting C++ compiler does not work as documentation would suggest

### ISSUE THREAD

**Issue by denisalevi** [brian-team/brian2#1322]
The documentation says:
> codegen.cpp.compiler = ''
>
>    Compiler to use (uses default if empty)
>
>    Should be gcc or msvc.

Setting
```
prefs.codegen.cpp.compiler = 'gcc'
```
fails with
```
/home/denis/.conda/envs/brian2-fork/lib/python3.8/site-packages/setuptools/distutils_patch.py:25: UserWarning: Distutils was imported before Setuptools. This usage is discouraged and may exhibit undesirable behaviors or errors. Please use Setuptools' objects directly or at least import Setuptools first.
  warnings.warn(
ERROR      Brian 2 encountered an unexpected error. If you think this is a bug in Brian 2, please report this issue either to the discourse forum at <http://brian.discourse.group/>, or to the issue tracker at <https://github.com/brian-team/brian2/issues>. Please include this file with debug information in your report: /tmp/brian_debug__rfpvc21.log  Additionally, you can also include a copy of the script that was run, available at: /tmp/brian_script_l8q0qgoh.py Thanks! [brian2]
Traceback (most recent call last):
  File "/home/denis/.conda/envs/brian2-fork/lib/python3.8/distutils/ccompiler.py", line 1008, in new_compiler
    (module_name, class_name, long_description) = compiler_class[compiler]
KeyError: 'gcc'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "example2_gchi_astrocyte.py", line 147, in <module>
    run(duration, report='text')
  File "/home/denis/tubcloud/home/projects/brian2cuda/brian2-fork/brian2/units/fundamentalunits.py", line 2430, in new_f
    result = f(*args, **kwds)
  File "/home/denis/tubcloud/home/projects/brian2cuda/brian2-fork/brian2/core/magic.py", line 373, in run
    return magic_network.run(duration, report=report, report_period=report_period,
  File "/home/denis/tubcloud/home/projects/brian2cuda/brian2-fork/brian2/core/magic.py", line 231, in run
    Network.run(self, duration, report=report, report_period=report_period,
  File "/home/denis/tubcloud/home/projects/brian2cuda/brian2-fork/brian2/core/base.py", line 277, in device_override_decorated_function
    return getattr(curdev, name)(*args, **kwds)
  File "/home/denis/tubcloud/home/projects/brian2cuda/brian2-fork/brian2/devices/cpp_standalone/device.py", line 1522, in network_run
    self.build(direct_call=False, **self.build_options)
  File "/home/denis/tubcloud/home/projects/brian2cuda/brian2-fork/brian2/devices/cpp_standalone/device.py", line 1146, in build
    compiler, default_extra_compile_args = get_compiler_and_args()
  File "/home/denis/tubcloud/home/projects/brian2cuda/brian2-fork/brian2/codegen/cpp_prefs.py", line 267, in get_compiler_and_args
    compiler_obj = new_compiler(compiler=compiler, verbose=0)
  File "/home/denis/.conda/envs/brian2-fork/lib/python3.8/distutils/ccompiler.py", line 1013, in new_compiler
    raise DistutilsPlatformError(msg)
distutils.errors.DistutilsPlatformError: don't know how to compile C/C++ code on platform 'posix' with 'gcc' compiler
```

I am running Arch Linux.

Looking into `distutils`, it seems to only know `unix` as compiler, not `gcc`. Maybe this is just a matter of changing the documentation? Or should `gcc` somehow work here?

And is there a preference (couldn't find one) or a way to specify the compiler binary directly? My arch linux just updated to gcc-11, which resulted in a bunch of errors of type 
```
/usr/bin/ld: synapses_pre_initialise_queue.cpp:(.text+0x17bd): undefined reference to `std::__throw_bad_array_new_length()'
```
Not sure if I have to manually rebuild some library, didn't figure it out yet (if you have an idea, please let me know).

Point is, I would have liked to use my `gcc10` binary instead, but didn't find a way to specify it. Is there a way? For now I just downgraded the `gcc` binary, but that is just a temporary workaround, I will want to upgrade it again soon.

**Comment by mstimberg** [comment]
This is indeed probably a bug in the documentation, we don't do anything with this preference ourselves, we just pass it on to distutils. We don't provide a way to specify the compiler binary directly, but you can set the `CXX` environment variable. Note that this was not completely working before merging #1304.

**Comment by denisalevi** [comment]
Ah right, `CXX` does what I want (after the merge you linked). Thanks. I opened a PR for the docs update.

**Comment by denisalevi** [comment]
For reference: The problem was that I updated my system `gcc`, but brian included the conda environment's `lib` instead of the system `lib`. Hence, the `libstd++` included was not the updated one, resulting in the `ld` error above.

On master, the issue does not persist. PR #1304 made sure that the conda environment's `gcc` is used instead of the system `gcc`, which should avoid out-of-sync libraries.

And it seems on the older brian version I was using, compiler tools are not installed with brian2 at all, so I didn't even have a conda environment specific `gcc`.

Closing this.
